# Warsztaty badawcze - Projekt 2

## Real Estate Price Analysis - Model AutoML z danymi geograficznymi zagregowanymi

### Wiktoria Boguszewska, Mateusz Zacharecki, Patrycja Żak


Uwaga: kod działa na colab lub lokalnie.

In [1]:
import warnings
warnings.filterwarnings('ignore')

In [2]:
!pip uninstall -y mkl
!pip install -U pip
!pip install -U "mxnet<2.0.0"
!pip install autogluon
!pip install -U ipykernel

!pip uninstall -y autogluon.core
!pip uninstall -y autogluon
!pip install autogluon
# po instalacji trzeba zrestartować środowisko

Found existing installation: mkl 2021.4.0
Uninstalling mkl-2021.4.0:
  Successfully uninstalled mkl-2021.4.0


[notice] A new release of pip is available: 24.2 -> 24.3.1
[notice] To update, run: python.exe -m pip install --upgrade pip
ERROR: To modify pip, please run the following command:
C:\Users\User\AppData\Local\Programs\Python\Python311\python.exe -m pip install -U pip



  Using cached pip-24.3.1-py3-none-any.whl.metadata (3.7 kB)
Using cached pip-24.3.1-py3-none-any.whl (1.8 MB)

  error: subprocess-exited-with-error
  
  × python setup.py bdist_wheel did not run successfully.
  │ exit code: 1
  ╰─> [264 lines of output]
      Running from numpy source directory.
      C:\Users\User\AppData\Local\Temp\pip-install-2ckfd5ew\numpy_0b8e9dea7901470a8580d392bbb02e94\numpy\distutils\misc_util.py:476: SyntaxWarning: "is" with a literal. Did you mean "=="?
        return is_string(s) and ('*' in s or '?' is s)
      blas_opt_info:
      blas_mkl_info:
      No module named 'numpy.distutils._msvccompiler' in numpy.distutils; trying from distutils
      customize MSVCCompiler
        libraries mkl_rt not found in ['C:\\Users\\User\\AppData\\Local\\Programs\\Python\\Python311\\lib', 'C:\\', 'C:\\Users\\User\\AppData\\Local\\Programs\\Python\\Python311\\libs']
        NOT AVAILABLE
      
      blis_info:
      No module named 'numpy.distutils._msvccompiler' in numpy.distutils; trying from distutils
      customize MSVCCompiler
        libraries blis not found in ['C:\\User


  Using cached mxnet-1.7.0.post2-py2.py3-none-win_amd64.whl.metadata (402 bytes)
  Using cached numpy-1.16.6.zip (5.1 MB)
  Preparing metadata (setup.py): started
  Preparing metadata (setup.py): finished with status 'done'
  Using cached requests-2.18.4-py2.py3-none-any.whl.metadata (50 kB)
  Using cached graphviz-0.8.4-py2.py3-none-any.whl.metadata (6.4 kB)
  Using cached chardet-3.0.4-py2.py3-none-any.whl.metadata (3.2 kB)
  Using cached idna-2.6-py2.py3-none-any.whl.metadata (8.9 kB)
  Using cached urllib3-1.22-py2.py3-none-any.whl.metadata (32 kB)
Using cached mxnet-1.7.0.post2-py2.py3-none-win_amd64.whl (33.1 MB)
Using cached graphviz-0.8.4-py2.py3-none-any.whl (16 kB)
Using cached requests-2.18.4-py2.py3-none-any.whl (88 kB)
Using cached chardet-3.0.4-py2.py3-none-any.whl (133 kB)
Using cached idna-2.6-py2.py3-none-any.whl (56 kB)
Using cached urllib3-1.22-py2.py3-none-any.whl (132 kB)
  Running setup.py clean for numpy
Failed to build numpy
  Using cached mkl-2021.4.0-py2.py3-

ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
torchaudio 2.4.1+cu121 requires torch==2.4.1+cu121, but you have torch 2.3.1 which is incompatible.

[notice] A new release of pip is available: 24.2 -> 24.3.1
[notice] To update, run: python.exe -m pip install --upgrade pip



Using cached mkl-2021.4.0-py2.py3-none-win_amd64.whl (228.5 MB)



[notice] A new release of pip is available: 24.2 -> 24.3.1
[notice] To update, run: python.exe -m pip install --upgrade pip


Found existing installation: autogluon.core 1.1.1
Uninstalling autogluon.core-1.1.1:
  Successfully uninstalled autogluon.core-1.1.1
Found existing installation: autogluon 1.1.1
Uninstalling autogluon-1.1.1:
  Successfully uninstalled autogluon-1.1.1
  Using cached autogluon-1.1.1-py3-none-any.whl.metadata (11 kB)
  Using cached autogluon.core-1.1.1-py3-none-any.whl.metadata (11 kB)
Using cached autogluon-1.1.1-py3-none-any.whl (9.7 kB)
Using cached autogluon.core-1.1.1-py3-none-any.whl (234 kB)



[notice] A new release of pip is available: 24.2 -> 24.3.1
[notice] To update, run: python.exe -m pip install --upgrade pip


In [40]:
from autogluon.tabular import TabularPredictor
from sklearn.model_selection import train_test_split
import pandas as pd

import random
random.seed(42)

import warnings
warnings.filterwarnings('ignore')

In [41]:
df = pd.read_csv('C:/Users/User/Desktop/STUDIA/Rok 2/Warsztaty badawcze/OLX/real_estate_imputed_geopy.csv')

In [42]:
df.head()

district_lon  district_lat  no_rooms          m  price_per_m    map_lon  \
0       16.9408      52.38356       3.0  48.000000      6020.83  16.959516   
1       16.9408      52.38356       3.0  61.000000      6213.11  16.950743   
2       16.9408      52.38356       3.0  63.000000      5380.95  16.945860   
3       16.9408      52.38356       3.0  59.200001      5574.32  16.980279   
4       16.9408      52.38356       2.0  38.000000      6815.79  16.950027   

     map_lat  building_floors_num  created_at_first_ts  created_at_ts  ...  \
0  52.391494                  4.0         1.515071e+09   1.515411e+09  ...   
1  52.380535                  4.0         1.516044e+09   1.516879e+09  ...   
2  52.393188                  4.0         1.515074e+09   1.515650e+09  ...   
3  52.391129                  4.0         1.515755e+09   1.517323e+09  ...   
4  52.395861                 10.0         1.514893e+09   1.516285e+09  ...   

   security_types                                    equipment_types  \
0            ['']                                               ['']   
1            ['']                                               ['']   
2  ['entryphone']                                               ['']   
3  ['entryphone']                                               ['']   
4  ['entryphone']  ['fridge', 'furniture', 'stove', 'washing_mach...   

                      extras_types     created_at_first           created_at  \
0  ['balcony', 'separate_kitchen']  2018-01-04 13:11:05  2018-01-08 11:27:35   
1                      ['balcony']  2018-01-15 19:12:07  2018-01-25 11:11:31   
2          ['balcony', 'basement']  2018-01-04 13:53:29  2018-01-11 05:55:12   
3          ['balcony', 'basement']  2018-01-12 11:10:57  2018-01-30 14:38:39   
4                     ['basement']  2018-01-02 11:34:51  2018-01-18 14:22:29   

   distance_to_park  distance_to_school  distance_to_bus_stop  \
0        597.265343          139.993206            240.462241   
1        365.846467          196.343583            143.404024   
2        490.321489          168.738469            199.005050   
3        262.952510          320.740826            163.058962   
4        391.986947          504.354945             62.490036   

   distance_to_market  distance_to_shopping_center  
0          125.901216                   615.599677  
1           75.995860                   321.301788  
2          191.345557                   365.353541  
3          356.891990                  1289.077156  
4          335.461563                   337.881818  

[5 rows x 92 columns]

In [43]:
X, y = df.loc[:, df.columns != "price_per_m"], df["price_per_m"]

In [44]:
def find_optimal_bins(X, y, max_bins=10, test_size=0.2, random_state=42):
    best_q = 1
    best_diff = float('inf')
    
    for q in range(2, max_bins + 1):
        y_binned = pd.qcut(y, q=q, duplicates='drop')  # tworzymy koszyki
        X_train, X_test, y_train, y_test = train_test_split(
            X, y, test_size=test_size, random_state=random_state, stratify=y_binned
        )
        # Sprawdzamy różnicę między średnimi
        diff = abs(y_train.mean() - y_test.mean())
        
        # Zapisujemy najlepszą liczbę koszyków, gdzie różnica jest najmniejsza
        if diff < best_diff:
            best_diff = diff
            best_q = q
    
    return best_q

In [45]:
optimal_bins = find_optimal_bins(X, y, max_bins=20)
print(optimal_bins)

15


In [46]:
y_binned = pd.qcut(y, q=optimal_bins, duplicates='drop')

In [47]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42, stratify=y_binned)

In [48]:
train_combined = pd.concat([X_train, y_train], axis=1)

In [ ]:
save_path = 'agModels-predictClass'
predictor = TabularPredictor(label='price_per_m', path=save_path, problem_type='regression', eval_metric='root_mean_squared_error').fit(train_combined, presets='best_quality')

Verbosity: 2 (Standard Logging)
=================== System Info ===================
AutoGluon Version:  1.1.1
Python Version:     3.11.3
Operating System:   Windows
Platform Machine:   AMD64
Platform Version:   10.0.19045
CPU Count:          8
Memory Avail:       6.43 GB / 15.37 GB (41.8%)
Disk Space Avail:   269.14 GB / 475.93 GB (56.5%)
Presets specified: ['best_quality']
Setting dynamic_stacking from 'auto' to True. Reason: Enable dynamic_stacking when use_bag_holdout is disabled. (use_bag_holdout=False)
Stack configuration (auto_stack=True): num_stack_levels=1, num_bag_folds=8, num_bag_sets=1
DyStack is enabled (dynamic_stacking=True). AutoGluon will try to determine whether the input data is affected by stacked overfitting and enable or disable stacking as a consequence.
	This is used to identify the optimal `num_stack_levels` value. Copies of AutoGluon will be fit on subsets of the data. Then holdout validation data is used to detect stacked overfitting.
	Running DyStack for up t

In [ ]:
performance = predictor.evaluate(train_combined)
print(performance)

{'root_mean_squared_error': -7207.567811211955, 'mean_squared_error': -51949033.753218696, 'mean_absolute_error': -504.4110537904304, 'r2': 0.1582536145854646, 'pearsonr': 0.6598565631723554, 'median_absolute_error': -295.076171875}


In [ ]:
predictor.fit_summary()

*** Summary of fit() ***
Estimated performance of each model:
                   model    score_val              eval_metric  pred_time_val    fit_time  pred_time_val_marginal  fit_time_marginal  stack_level  can_infer  fit_order
0    WeightedEnsemble_L3 -7783.395859  root_mean_squared_error     222.149774  794.571416                0.003000           0.142127            3       True          6
1    WeightedEnsemble_L2 -7785.269224  root_mean_squared_error     109.728972  511.537135                0.003997           0.167037            2       True          4
2      LightGBMXT_BAG_L1 -7785.768103  root_mean_squared_error       7.466338  486.725680                7.466338         486.725680            1       True          3
3      LightGBMXT_BAG_L2 -7794.846181  root_mean_squared_error     222.146774  794.429290                6.840101         257.924922            2       True          5
4  KNeighborsUnif_BAG_L1 -8268.038179  root_mean_squared_error     102.258638   24.644418         

{'model_types': {'KNeighborsUnif_BAG_L1': 'StackerEnsembleModel_KNN',
  'KNeighborsDist_BAG_L1': 'StackerEnsembleModel_KNN',
  'LightGBMXT_BAG_L1': 'StackerEnsembleModel_LGB',
  'WeightedEnsemble_L2': 'WeightedEnsembleModel',
  'LightGBMXT_BAG_L2': 'StackerEnsembleModel_LGB',
  'WeightedEnsemble_L3': 'WeightedEnsembleModel'},
 'model_performance': {'KNeighborsUnif_BAG_L1': -8268.03817892719,
  'KNeighborsDist_BAG_L1': -9250.642986539868,
  'LightGBMXT_BAG_L1': -7785.76810287451,
  'WeightedEnsemble_L2': -7785.269223829596,
  'LightGBMXT_BAG_L2': -7794.846181164716,
  'WeightedEnsemble_L3': -7783.395858800736},
 'model_best': 'WeightedEnsemble_L3',
 'model_paths': {'KNeighborsUnif_BAG_L1': ['KNeighborsUnif_BAG_L1'],
  'KNeighborsDist_BAG_L1': ['KNeighborsDist_BAG_L1'],
  'LightGBMXT_BAG_L1': ['LightGBMXT_BAG_L1'],
  'WeightedEnsemble_L2': ['WeightedEnsemble_L2'],
  'LightGBMXT_BAG_L2': ['LightGBMXT_BAG_L2'],
  'WeightedEnsemble_L3': ['WeightedEnsemble_L3']},
 'model_fit_times': {'KNeigh

In [ ]:
test_combined = pd.concat([X_test, y_test], axis=1)

In [ ]:
rmse_score = predictor.evaluate(test_combined)
print("RMSE na zbiorze testowym:", rmse_score)

RMSE na zbiorze testowym: {'root_mean_squared_error': -798.7603277417487, 'mean_squared_error': -638018.0611741059, 'mean_absolute_error': -466.6535873058953, 'r2': 0.558276542291136, 'pearsonr': 0.7507827554926154, 'median_absolute_error': -311.22998046875}


In [ ]:
print("RMSE:", round(abs(rmse_score['root_mean_squared_error']),2))

RMSE: 798.76


RMSE: 635.5

Na zbiorze real_estate_imputed osiągnięto w poprzedniej iteracji RMSE 635.5.
Na zbiorze z danymi surowymi z dodanymi kolumnami geograficznymi osiągnięto 798.76.
Chcieliśmy wywołać poprzedni kod jeszcze raz, ale ewakuacja na uczelni zepsuła nam plany.